### PR TITLE
Modernize developer guide and nix-shell code

### PR DESCRIPTION
**Pull request by sdht0**
* Update CLion -> RustRover
* Add info about git submodules
* Mention more dependencies for lint checks
* Update nix-shell code. Tested on both Linux and macOS. 

### Tips for reviewer

* I removed the RUSTFLAGS that were causing errors.
* I'll rebase the commits after addressing any feedback.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: doc/developer/guide.md [hunk 1]
```diff
@@ -180,6 +179,26 @@ location, set `$CONFLUENT_HOME` to this location and add `$CONFLUENT_HOME/bin`
...
+Optionally, you can use [nix][Nix] to install all required dependencies on both Linux and macOS,
+using the provided [`shell.nix`](../../misc/nix). [Install nix][nix-manual] and use `nix-shell` to enter an
+environment that is isolated from the main OS.
+
+```bash
+nix-shell misc/nix/shell.nix
+[nix-shell]$ rustup install stable # If not installed already
+```
+
+Materialize can then be build inside this shell. Note that CockroachDB is not included in the above configuration
```
Comment (benesch) [inline doc/developer/guide.md:193]:
```suggestion
Materialize can then be built inside this shell. Note that CockroachDB is not included in the above configuration
```

File: doc/developer/guide.md [hunk 2]
```diff
@@ -192,6 +211,13 @@ Because the MaterializeInc organization requires two-factor authentication
 (2FA), you'll need to clone via SSH as indicated above, or [configure a personal
 access token for use with HTTPS][github-https].
 
+Optionally, you may need to also clone the associated submodules:
+
+```shell
+git submodule init
+git submodule update --recursive
```
Comment (benesch) [inline doc/developer/guide.md:218]:
```suggestion
git submodule update --init --recursive
```

A bit shorter!

File: misc/nix/shell.nix [hunk 3]
```diff
@@ -10,33 +10,30 @@
...
+    darwin.apple_sdk.frameworks.DiskArbitration
+    darwin.apple_sdk.frameworks.Foundation
+  ];
 
-  shellHook = ''
-    export LD_LIBRARY_PATH=${pkgs.stdenv.cc.cc.lib}/lib:$LD_LIBRARY_PATH
-    export LIBCLANG_PATH="${llvmPackages_14.libclang.lib}/lib";
-  '';
+  RUSTFLAGS = "-Cdebuginfo=1 -Csymbol-mangling-version=v0";
 }
```
Comment (benesch) [inline misc/nix/shell.nix:39]: These changes seem material. Can you check with @petrosagg that they won't break his workflow? 

Then again the `--cfg=tokio-usntable` typo doesn't exactly inspire confidence that was was here was working well...

**Review by benesch — COMMENTED** [review]
Changes to the developer guide LGTM, thanks! But please get someone who uses the Nix stuff—probably @petrosagg—to sign off on the Nix changes before merging! (Or split into two PRs if you want to merge the docs stuff sooner.)

File: misc/nix/shell.nix (hunk 3 above)
Comment (petrosagg) [reply misc/nix/shell.nix:39]: I'll try this out but is there a reason we're removing `mold` as the linker?

Comment (sdht0) [reply misc/nix/shell.nix:39]: `mold` doesn't work on macOS, so I had made it conditional for Linux only at first, but even there I was getting errors. I can take another look if we want to keep mold around on Linux only.

Same for the other flags, which I removed one by one until the shell worked.

Comment (petrosagg) [reply misc/nix/shell.nix:39]: I looked into this today I get some linking errors when optionally enabling `mold` for Linux. I'll have to check which of the changes breaks it. The failure I got was failing to find `Scrt1.o`

Comment (sdht0) [reply misc/nix/shell.nix:39]: Does the original code work for you? `bin/environmentd` fails for me while building `rdkafka-sys` with openssl related errors.

Comment (petrosagg) [reply misc/nix/shell.nix:39]: It works, yeah. I only have issues with building the `openssl` crate when it's built in non-vendored mode but we do have the `vendored` feature enabled in MZ. Maybe that is ignored on OSX?

Comment (petrosagg) [reply misc/nix/shell.nix:39]: The new code here also works for me, I tried a build from scratch, I just wanted to keep the mold option for linux

Comment (sdht0) [reply misc/nix/shell.nix:39]: > It works, yeah. I only have issues with building the `openssl` crate when it's built in non-vendored mode but we do have the `vendored` feature enabled in MZ. Maybe that is ignored on OSX?

Ah okay. I tried on Linux actually. With the original shell.nix, I get:
```bash
$ nix-shell misc/nix/shell.nix
$ bin/environmentd
  Building and linking librdkafka statically
  CMake Deprecation Warning at CMakeLists.txt:1 (cmake_minimum_required):
    Compatibility with CMake < 3.5 will be removed from a future version of
    CMake.

    Update the VERSION argument <min> value or use a ...<max> suffix to tell
    CMake that the project does not need compatibility with older versions.

  CMake Error at src/CMakeLists.txt:235 (get_target_property):
    get_target_property() called with non-existent target "OpenSSL::SSL".

  thread 'main' panicked at /home/artimaeus/.cargo/registry/src/index.crates.io-6f17d22bba15001f/cmake-0.1.50/src/lib.rs:1098:5:
```
Am I missing something?

**Review by petrosagg — APPROVED** [review]

File: misc/nix/shell.nix (hunk 3 above)
Comment (sdht0) [reply misc/nix/shell.nix:41]: Tested on both macOS and Linux. `nix-shell --pure misc/nix/shell.nix` successfully runs `bin/environmentd`.

**Pull request merged**